### PR TITLE
System variables

### PR DESCRIPTION
**Pull request by fenilfadadu**

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/myadapter/command/commands/QueryMessageProcessor.java [hunk 1]
```diff
@@ -115,30 +113,15 @@ public void processMessage(WireMessage message) throws Exception {
...
-    backendConnection.setAutocommit(false);
-  }
 
   private void processResultSet(ResultSet resultSet, QueryReplacement queryReplacement)
       throws Exception {
-    sendColumnDefinitions(resultSet, queryReplacement);
+    boolean shouldSendColumnDefinitions = true;
     while (resultSet.next()) {
+      if (shouldSendColumnDefinitions) {
+        sendColumnDefinitions(resultSet, queryReplacement);
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/command/commands/QueryMessageProcessor.java:135]: Column definition need to be sent irrespective of where there is row or not.

File: src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java [hunk 2]
```diff
@@ -16,14 +16,18 @@
...
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
 
 /**
  * Parse specified data to boolean. For most cases it is simply returning the bit representation.
  */
 @InternalApi
 public class BooleanParser extends Parser<Boolean> {
-  private static final byte[] TRUE_VALUE_BYTES = new byte[] {'1'};
+  static final byte[] TRUE_VALUE_BYTES = new byte[] {'1'};
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java]: Why is this required?

File: src/main/java/com/google/cloud/spanner/myadapter/ConnectionHandler.java [hunk 3]
```diff
@@ -90,10 +90,10 @@ public WireProtocolHandler getWireHandler() {
             String.format(
                 "Connection handler with ID %s created for client %s",
                 getName(), socket.getInetAddress().getHostAddress()));
-    this.sessionState = new SessionState();
     this.options = server.getOptions();
     this.backendConnection =
         new BackendConnection(options, server.getProperties(), spannerConnection);
+    this.sessionState = new SessionState(server.getOptions(), backendConnection);
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/ConnectionHandler.java]: I don't think SessionState should be accessing backendConnection.

File: src/main/java/com/google/cloud/spanner/connection/BackendConnection.java [hunk 4]
```diff
@@ -111,8 +116,14 @@ public void connectToSpanner(String database, @Nullable Credentials credentials)
     this.databaseId = connectionOptions.getDatabaseId();
   }
 
-  public StatementResult executeQuery(Statement originalStatement) {
-    return spannerConnection.execute(originalStatement);
+  public StatementResult executeQuery(
+      Statement statement, ParsedStatement parsedStatement, SessionState sessionState) {
+    SessionStatement sessionStatement = SessionStatementParser.parse(parsedStatement);
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/connection/BackendConnection.java]: Can we move the parsing to QueryMessageProcessor and only keep execution here?

File: src/main/java/com/google/cloud/spanner/connection/BackendConnection.java [hunk 5]
```diff
@@ -145,4 +156,77 @@ private String appendPropertiesToUrl(String url, Properties info) {
     }
     return result.toString();
   }
+
+  @InternalApi
+  public static final class QueryResult implements StatementResult {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/connection/BackendConnection.java]: Create separate files for both the classes below.

File: src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java [hunk 6]
```diff
@@ -16,14 +16,18 @@
...
   private static final byte[] FALSE_VALUE_BYTES = new byte[] {'0'};
+  public static final Set<String> TRUE_VALUES = ImmutableSet.of("true", "1");
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java:29]: Same

File: src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java [hunk 7]
```diff
@@ -16,14 +16,18 @@
...
+  public static final Set<String> FALSE_VALUES = ImmutableSet.of("false", "0");
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java:30]: Same.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 8]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+    internalSet(name.toLowerCase(Locale.ROOT), value, variableMap);
+  }
+
+  private void internalSet(String name, String value, Map<String, SystemVariable> variableMap) {
+    if ("names".equals(name)) {
+      // TODO: Consider handling "SET NAMES" as a separate statement type.
+      handleNames(value);
+      return;
+    }
+    if ("autocommit".equals(name)) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Move this to a static variable.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 9]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+  public void set(String name, String value, SessionVariableScope scope) {
+    logger.log(
+        Level.INFO,
+        () -> String.format("Setting system variable %s to %s at scope %s", name, value, scope));
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(scope);
+    internalSet(name.toLowerCase(Locale.ROOT), value, variableMap);
+  }
+
+  private void internalSet(String name, String value, Map<String, SystemVariable> variableMap) {
+    if ("names".equals(name)) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Static variable.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 10]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+    if (variable == null) {
+      throw unknownVariableError(name);
+    }
+    variable.setValue(value);
+  }
+
+  private String handleAutocommit(String value) {
+    if (BooleanParser.TRUE_VALUES.contains(value)) {
+      backendConnection.processSetAutocommit();
+      return "1";
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Static variable.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 11]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+    }
+    if (BooleanParser.FALSE_VALUES.contains(value)) {
+      backendConnection.processUnsetAutocommit();
+      return "0";
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Static variable.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 12]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+    }
+    throw invalidValueError("autocommit", value);
+  }
+
+  private void handleNames(String value) {
+    logger.log(Level.INFO, () -> String.format("Setting all character sets to %s", value));
+    List<String> charasets =
+        ImmutableList.of(
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Static variable.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 13]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+  /** Returns the current value of the specified setting. */
+  public SystemVariable get(String name, SessionVariableScope scope) {
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(scope);
+    return internalGet(name.toLowerCase(Locale.ROOT), variableMap);
+  }
+
+  private SystemVariable internalGet(String key, Map<String, SystemVariable> variableMap) {
+    SystemVariable variable = variableMap.get(key);
+    if (variable == null) {
+      throw unknownVariableError(key);
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java:155]: Is this inline with MySql?

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 14]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+    SystemVariable variable = variableMap.get(name);
+    if (variable == null) {
+      throw unknownVariableError(name);
+    }
+    variable.setValue(value);
+  }
+
+  private String handleAutocommit(String value) {
+    if (BooleanParser.TRUE_VALUES.contains(value)) {
+      backendConnection.processSetAutocommit();
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Instead on SessionState calling backend connection, design the method to return the operation. Otherwise session state will become very bloted. Also, this introduces circular dependency which is an Anti-pattern.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SystemVariable.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+import javax.annotation.Nonnull;
+
+/** Represents a row in the pg_settings table. */
+@InternalApi
+public class SystemVariable {
+  private static final int NAME_INDEX = 0;
+  private static final int DATATYPE_INDEX = 1;
+  private static final int VALUE_INDEX = 2;
+
+  private String extension;
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SystemVariable.java]: What is the role of this?

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+    SetStatement(Builder builder) {
+      this.systemVariable = builder.systemVariable;
+      this.scope = builder.scope;
+      this.name = builder.name;
+      this.value = builder.value;
+    }
+
+    @Override
+    public StatementResult execute(SessionState sessionState) {
+      sessionState.set(name, value, scope);
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Consider binging the autocommit logic here. This seems a more reasonable place. Although make sure that Spanner is updated first.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+  }
+
+  public static @Nullable SessionStatement parse(ParsedStatement parsedStatement) {
+    if (parsedStatement.getType() == StatementType.CLIENT_SIDE) {
+      // TODO : See if we need special handling for client side statements.
+      // Disabling this right now as "SET AUTOCOMMIT 1" is getting treated as
+      // client side statement, but it's not a valid GoogleSQL statement.
+    }
+    SimpleParser parser = new SimpleParser(parsedStatement.getSqlWithoutComments());
+    if (parser.eatKeyword("set")) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Validate is SET also works. Also use static variables.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+      return parseSetStatement(parser);
+    }
+    if (parser.eatKeyword("select")
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Static strings.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+    return null;
+  }
+
+  static SetStatement parseSetStatement(SimpleParser parser) {
+    SetStatement.Builder builder = new SetStatement.Builder();
+    if (parser.eatKeyword("names")) {
+      builder.systemVariable();
+      builder.name("names");
+    } else {
+      if (parser.eatToken("@@") || !parser.eatToken("@")) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: What is the need for second check?

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+        && parser.peek(true, false, "@@")) {
+      return parseSelectStatement(parser);
+    }
+
+    return null;
+  }
+
+  static SetStatement parseSetStatement(SimpleParser parser) {
+    SetStatement.Builder builder = new SetStatement.Builder();
+    if (parser.eatKeyword("names")) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Validate for upper case.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java [hunk 21]
```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
...
+        return settings;
+      default:
+        throw unknownParamError("scope");
+    }
+  }
+  /**
+   * Sets the value of the specified setting. The new value will be persisted if the current
+   * transaction is committed. The value will be lost if the transaction is rolled back.
+   */
+  public void set(String name, String value, SessionVariableScope scope) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Convert names to either lower or upper case to ensure both the cases are covered.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+      TableOrIndexName name = parser.readTableOrIndexName();
+      if (name == null) {
+        throw SpannerExceptionFactory.newSpannerException(
+            ErrorCode.INVALID_ARGUMENT,
+            "Invalid SET statement: "
+                + parser.getSql()
+                + ". Expected configuration parameter name.");
+      }
+
+      if ("global".equals(name.getUnquotedSchema())) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Static

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+    if (parser.eatKeyword("select")
+        && !parser.peek(true, true, "FROM")
+        && parser.peek(true, false, "@@")) {
+      return parseSelectStatement(parser);
+    }
+
+    return null;
+  }
+
+  static SetStatement parseSetStatement(SimpleParser parser) {
```
Comment (pratickchokhani) [inline src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java:221]: Can there be multiple variables be set in a single query?

File: src/main/java/com/google/cloud/spanner/myadapter/command/commands/QueryMessageProcessor.java (hunk 1 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/command/commands/QueryMessageProcessor.java:135]: As I said in the call also, sendColumnDefinitions accesses the ResultSet (particularly calls getType). But for pre-populated ResultSets .next() must be called at least once before accessing ResultSetMethods. I have already enhanced this code block, please checkout diff WRT latest commits.

File: src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java (hunk 2 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java]: Not needed. Will undo.

File: src/main/java/com/google/cloud/spanner/myadapter/ConnectionHandler.java (hunk 3 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/ConnectionHandler.java]: This is needed for autocommit handling. Whenever autocommit system variable is modified, we need to propagate that to Spanner via backendConnection. So this looked like a clear dependency. Couldn't think of a cleaner approach to do that.

File: src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java (hunk 7 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/parsers/BooleanParser.java:30]: This is public as we use it to determine if the received value is true or false even outside this class. For example autocommit can be set by either setting it to 'true' or to '1'.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java (hunk 13 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java:155]: Yes, although the error message shown was not completely consistent. I'll fix it in the latest commit.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java (hunk 23 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java:221]: Not mentioned in the documentation, but based on experimentation, yes. Although I feel this is not used much based on exploration. We can have a TODO for now, and we probably don't need to support it even in our first release.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SystemVariable.java (hunk 15 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/session/SystemVariable.java]: This got copied over from PGAdapter, but is not needed now. I mentioned in our last conversation that I'll remove it as part of the use defined variable implementation, but let me go ahead and clean up now itself to avoid confusion.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java (hunk 17 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Yes SimpleParser methods like eatKeyWord and eatToken ignores the case. So both SET and set works. Here as well as other places also.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java (hunk 19 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: SET command accepts system variables either with @@ prefix, or even with no prefix at all, but @ prefix means user defined variable.

File: src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java (hunk 20 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/statements/SessionStatementParser.java]: Yes, SimpleParser itself supports it.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java (hunk 21 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: Already noticed it and have fixed it.

File: src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java (hunk 14 above)
Comment (fenilfadadu) [reply src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java]: I get your point, but I couldn't think of a generic way of "returning the operation" here.  So I have tried to solved it using a patten of having a pre and post hooks(or handlers) for an operation. Please check the latest commit addressing the comments.

**Comment by fenilfadadu** [comment]
Addressed Pratick's comments.

**Review by pratickchokhani — APPROVED** [review]
